src/ArgumentEscaper.cs

```diff
@@ -0,0 +1,207 @@
+        var isQuoted = needsQuotes || IsSurroundedWithQuotes(arg);
+
+        if (needsQuotes)
+        {
+            sb.Append(Quote);
+        }
+
+        for (var i = 0; i < length; ++i)
+        {
+            var backslashCount = 0;
+
+            // Consume All Backslashes
+            while (i < arg.Length && arg[i] == Backslash)
+            {
+                backslashCount++;
+                i++;
+            }
+
+            // Escape any backslashes at the end of the arg
+            // when the argument is also quoted.
+            // This ensures the outside quote is interpreted as
+            // an argument delimiter
+            if (i == arg.Length && isQuoted)
+            {
+                sb.Append(Backslash, 2 * backslashCount);
+            }
+
+            // At then end of the arg, which isn't quoted,
+            // just add the backslashes, no need to escape
+            else if (i == arg.Length)
+            {
+                sb.Append(Backslash, backslashCount);
+            }
+
+            // Escape any preceding backslashes and the quote
+            else if (arg[i] == Quote)
+            {
+                sb.Append(Backslash, (2 * backslashCount) + 1);
+                sb.Append(Quote);
+            }
+
+            // Output any consumed backslashes and the character
+            else
+            {
+                sb.Append(Backslash, backslashCount);
+                sb.Append(arg[i]);
+            }
+        }
+
+        if (needsQuotes)
+        {
+            sb.Append(Quote);
+        }
+    }
+
+    /// <summary>
+    /// <para>Prepare as single argument to roundtrip properly through cmd.</para>
+    /// <para>This prefixes every character with the <c>^</c> character to force cmd to interpret the argument string literally.</para>
+    /// </summary>
+    /// <remarks>
+    /// See here for more info: <seealso href="https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way"/>.
+    /// </remarks>
+    /// <param name="sb">The <seealso cref="ValueStringBuilder"/> to append to.</param>
+    /// <param name="argument">The argument to escape.</param>
+    private static void EscapeArgForCmd(ref ValueStringBuilder sb, string argument)
+    {
+        var quoted = ArgumentContainsWhitespace(argument);
+
+        if (quoted)
+        {
+            sb.Append(Caret);
+            sb.Append(Quote);
+        }
+
+        // Prepend every character with ^
+        // This is harmless when passing through cmd
+        // and ensures cmd metacharacters are not interpreted
+        // as such
+        foreach (var character in argument)
+        {
+            sb.Append(Caret);
+            sb.Append(character);
+        }
+
+        if (quoted)
+        {
+            sb.Append(Caret);
+            sb.Append(Quote);
+        }
+    }
+
+    private static bool IsSurroundedWithQuotes(ReadOnlySpan<char> argument)
+        => argument[0] == Quote &&
```

Comment:
May be worth adding an `argument.Length >= 2` check

Reply:
I don't think it's possible to get a value here that's just a single quote. If there is I haven't been able to reproduce it via the command line. Single character arguments are fine but probably aren't worth adding another check when most are going to be multiple characters.